### PR TITLE
Assembly Line Scheduling

### LINKED ISSUE #774: Assembly Line Scheduling
Language Implementation: 
- [x] C
- [x] C++
- [x] Python
- [x] Java
- [x] Php
- [x] C#

### PR DESCRIPTION
**Pull request by Karna98**
Fixes #774 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [X] Logic Documentation (Comments).
- [X] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Added Implementation of Assembly Line Scheduling

#### Languages Used:
- C
- C++
- C#
- Java

#### Files Added:
- Assembly_Line_Scheduling/Assembly_Line_Scheduling.c
- Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp
- Assembly_Line_Scheduling/Assembly_Line_Scheduling.cs
- Assembly_Line_Scheduling/Assembly_Line_Scheduling.java

Thanks!

### CODE REVIEW

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.c [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+#define assembly_lines 2
+#define nodes 3
+
+int ALS(int line[][nodes], int in_time[], int out_time[], int t_time[][nodes], int path[])
+{
+    // line 	-> containing assembly line with its node values
+    // t_time 	-> containing lists of Extra cost incurred to change the assembly line
+    // in_time 	-> list of values of In_time for assembly lines
+    // out_time -> list of values of Out_time for assembly lines
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.c]: Remove the tabs in these lines

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.c [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        out_time[0] = out_a	 		and 	out_time[1] = out_b 
+    */
+
+    //Example1
+    
+    int line	[assembly_lines][nodes] = { { 80, 40, 60 }, { 50, 70, 50 } };
+    int in_time	[assembly_lines] 		= { 30, 10 }; 
+    int out_time[assembly_lines] 		= { 30, 30 };
+    int t_time	[assembly_lines][nodes] = { { 0, 20, 20 }, { 0, 10, 30 } };
+    int path	[nodes+2];
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.c]: Remove the tabs in above lines

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.c [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    
+    //Example2
+    /*
+    int line	[assembly_lines][nodes] = { { 4, 5, 3, 2 }, { 2, 10, 1, 4 } };
+    int in_time	[assembly_lines] 		= { 10, 12 }; 
+    int out_time[assembly_lines] 		= { 18, 7 };
+    int t_time	[assembly_lines][nodes] = { { 0, 7, 4, 5 }, { 0, 9, 2, 8 } };
+    int path	[nodes+2];
+    */
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.c]: Remove the tabs in above lines

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    tAi,j -> Translating from Ai to Bj
+    tBi,j -> Translating from Bi to Aj
+    j = i + 1 for most of cases.
+*/
+
+#include<iostream>
+using namespace std;
+
+#define assembly_lines 	2
+#define nodes 			3
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp]: Remove tabs

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+int ALS( int line[][nodes], int in_time[], int out_time[], int t_time[][nodes], int path[] )
+{
+    // line 	-> containing assembly line with its node values
+    // t_time 	-> containing lists of Extra cost incurred to change the assembly line
+    // in_time 	-> list of values of In_time for assembly lines
+    // out_time -> list of values of Out_time for assembly lines
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp]: Remove tabs

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        
+        line_a, line_b time required at each node on assembly line
+        a_to_b, b_to_a time required at change assembly line from Ai to Bj or Bi to Aj where j = i+1
+        in_a, in_b Time required to enter Assembly line
+        out_a, out_b Time required to exit from Assembly line
+
+        line[0] 	= line_a 		and 	line[1] 	= line_b
+        t_time[0] 	= a_to_b 		and 	t_time[1] 	= b_to_a
+        in_time[0] 	= in_a 			and 	in_time[1] 	= in_b 
+        out_time[0] = out_a	 		and 	out_time[1] = out_b
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp]: Replace tabs with spaces

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        out_time[0] = out_a	 		and 	out_time[1] = out_b 
+    */
+
+    //Example1
+    
+    int line	[assembly_lines][nodes] = { { 80, 40, 60 }, { 50, 70, 50 } };
+    int in_time	[assembly_lines] 		= { 30, 10 }; 
+    int out_time[assembly_lines] 		= { 30, 30 };
+    int t_time	[assembly_lines][nodes] = { { 0, 20, 20 }, { 0, 10, 30 } };
+    int path	[nodes+2];
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp]: Remove tabs

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    
+
+    //Example2
+    /*
+    int line	[assembly_lines][nodes] = { { 4, 5, 3, 2 }, { 2, 10, 1, 4 } };
+    int in_time	[assembly_lines] 		= { 10, 12 }; 
+    int out_time[assembly_lines] 		= { 18, 7 };
+    int t_time	[assembly_lines][nodes] = { { 0, 7, 4, 5 }, { 0, 9, 2, 8 } };
+    int path	[nodes+2];
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp]: Remove tabs.

**Comment by Karna98** [comment]
> You have to correct these:
> 
>     1. Remove the space before `,`
> 
>     2. Add a space after `,` and `;`
> 
>     3. Remove tabs where you have used them in between some statements and comments. Either remove them or replace them with spaces. (I have mentioned this mistake in c and cpp files. Correct the others also)

Updated. Please check the latest commit.

**Comment by Karna98** [comment]
> You had to correct the spacing before and after comma and semi-colon also.

Hope it looks good to merge.

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.c [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        temp_c = line1[node] + line[1][node] + t_time[0][node];
+        temp = min(temp_s, temp_c);
+        line2[node+1] = temp;
+    }
+
+    // Adding Out time for respective line
+    line1[nodes+1] = line1[nodes] + out_time[0];
+    line2[nodes+1] = line2[nodes] + out_time[1];
+
+    int min_line = min(line1[nodes+1],line2[nodes+1]);
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.c]: Add a space after `,`

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+    int val = 0;
+    val = ALS(line, in_time, out_time, t_time, path);
+ 
+    cout << "Minimum Time : " << val << "\n";
+    for (int node = 0; node < (nodes + 2); node++)
+    {
+        if(node == 0)
+        {
+            cout << "Path : In" << (path[node]+1) << "(" << in_time[path[node]] << ") --> ";
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp]: Remove the space before `:`

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.php [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+            {
+                array_splice( $path,sizeof($path), 0, 'Out'.$path_c.'('.$out_time[(int)$path_c-1].')');
+            }
+            elseif ($i == 0)
+            {
+                array_splice( $path,0, 0, 'In'.$path_c.'('.$in_time[(int)$path_c-1].')');
+            }
+            else
+            {
+                array_splice( $path,0, 0, 'Node'.$path_c.' ('.$line[(int)$path_c-1][$i-1].')');
```
Comment (SakshiShreya) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.php]: Add a space after `$path,`

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp (hunk 10 above)
Comment (Karna98) [reply Assembly_Line_Scheduling/Assembly_Line_Scheduling.cpp]: That is just for output can look good.

**Comment by Karna98** [comment]
> Sorry, I did not notice this earlier but you have not followed naming conventions.
> 
>     * Use camelCase for variable and function names.
> 
>     * Use PascalCase for class names.
> 
>     * Use UNDERSCORE_SEPARATED_UPPERCASE for constant names (the ones after #define)

If class names are changed then I have to change the file name for it (ex java file)
So File name conventions are A_B_C but it will become something like ABC. Is that okay?

**Comment by Karna98** [comment]
Pushed the latest commit. If anything more required let me know. Hope its good to go.

**Comment by Karna98** [comment]
- [x] Done with renaming functions according to camelCase.

**Comment by jainaman224** [comment]
@Karna98 Great work
Please break this PR into 2 if you want to get more marks as there is a capping of 50 marks for each PR.
Please take input dynamically.

**Comment by Karna98** [comment]
> @Karna98 update?

Pushing Php and Python through different PR.
Added Dynamic Input to C, C++ ,C# and Java

**Comment by Karna98** [comment]
Please check the latest commit. Required changes are committed.

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.c [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+    {
+        printf("For Line %d : ", lineN + 1);
+        scanf("%d", &outTime[lineN]);
+    }
+    
+    printf("\n-- Time required at each node --\n");
+    for (int lineN = 0; lineN < ASSEMBLY_LINES; lineN++)
+    {
+        printf("\nFor Assembly Line %d\n", lineN+1);
+        for(int nThNode = 0; nThNode < nodes; nThNode++)
```
Comment (MastersAbh) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.c]: space after for

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.cs [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+                }
+                else
+                {
+                    int val_t = line1[node] - line1[node - 1];
+                    if (val_t != line[pathC, node - 1])
+                    {
+                        pathC = 1;
+                    }
+                }
+
```
Comment (MastersAbh) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.cs]: please remove this extra line

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.java [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+
+        System.out.print("Enter the no. of nodes on Assembly line : ");
+        Scanner s = new Scanner(System.in);
+        nodes = s.nextInt();
+
+        int line   [][]    = new int[assemblyLines][nodes];
+        int inTime []      = new int[assemblyLines];
+        int outTime[]      = new int[assemblyLines];
+        int tTime  [][]    = new int[assemblyLines][nodes];
+        int path   []      = new int[nodes + 2];
```
Comment (MastersAbh) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.java]: please remove extra spaces before =. There should be 1 space

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.c [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+    tAi,j -> Translating from Ai to Bj
+    tBi,j -> Translating from Bi to Aj
+    j = i + 1 for most of cases.
+*/
+
+#include<stdio.h>
+
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+
+#define ASSEMBLY_LINES   2
```
Comment (MastersAbh) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.c]: give only one space before 2

**Comment by Karna98** [comment]
Updated Changes.

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.cs [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+        for (int lineN = 0; lineN < assemblyLines; lineN++)
+        {
+            Console.Write("For Line" + (lineN + 1) +": ");
+            inTime[lineN] = Convert.ToInt32(Console.ReadLine());
+        }
+
+        Console.WriteLine("\n-- Out Time --\n");
+        for (int lineN = 0; lineN < assemblyLines; lineN++)
+        {
+            Console.Write("For Line" + (lineN + 1) +": ");
```
Comment (MastersAbh) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.cs]: please add space after +

File: Assembly_Line_Scheduling/Assembly_Line_Scheduling.java [hunk 17]
```diff
@@ -0,0 +1,282 @@
...
+        */
+
+        int assemblyLines, nodes;
+        assemblyLines  = 2;
+
+        System.out.print("Enter the no. of nodes on Assembly line : ");
+        Scanner s = new Scanner(System.in);
+        nodes = s.nextInt();
+
+        int line   [][] = new int[assemblyLines][nodes];
```
Comment (MastersAbh) [inline Assembly_Line_Scheduling/Assembly_Line_Scheduling.java]: remove spaces before [] through line 167 to 171

**Comment by marwahmanbir** [comment]
@Karna98 also do squash your commits after making the final changes.

**Comment by Karna98** [comment]
@MastersAbh Requested changes are Done.
@marwahmanbir Squashed Commits.

**Pull request merged**